pkg/cache/cluster.go

```diff
@@ -365,22 +362,20 @@ func runSynced(lock sync.Locker, action func() error) error {
 	return action()
 }
 
-func (c *clusterCache) watchEvents(ctx context.Context, api kube.APIResourceInfo, info *apiMeta, resClient dynamic.ResourceInterface, ns string) {
+func (c *clusterCache) watchEvents(ctx context.Context, api kube.APIResourceInfo, resClient dynamic.ResourceInterface, ns string, resourceVersion string) {
 	kube.RetryUntilSucceed(ctx, watchResourcesRetryTimeout, fmt.Sprintf("watch %s on %s", api.GroupKind, c.config.Host), func() (err error) {
 		defer func() {
 			if r := recover(); r != nil {
 				err = fmt.Errorf("Recovered from panic: %+v\n%s", r, debug.Stack())
 			}
 		}()
 
-		err = runSynced(&c.lock, func() error {
-			if info.resourceVersion != "" {
-				return nil
-			}
+		// load API initial state if no resource version provided
+		if resourceVersion == "" {
```

Comment:
Inner block of this `if` should be in a closure/function to ensure `defer c.listSemaphore.Release(1)` is executed ASAP. Otherwise the semaphore will be held for the duration of the watch which is obviously not what we want. Previously, as you can see, we relied on `runSynced()` closure.

Reply:
Thank you! Nice catch. Fixed

Reply:
I think the fix is not good enough :) For two reasons:
- Previously the semaphore was held during all the lifetime of the `items` variable. Remember, we want to limit memory pressure so it should not only cover the pager but the whole section where `items` variable is used. I think this is important.
- Pager may panic but we recover from panics above. This will lead to semaphore permits leakage. So I think it really should be a closure + `defer` inside. And also I'd get rid of all the panic recover code from everywhere as this cure may be worse than the disease.

Reply:
Thank you @ash2k ! Created follow up PR to address it: https://github.com/argoproj/gitops-engine/pull/135